# Jane Street Puzzles - May 2020

A solution by Walter Sebastian Gisler

## Problem Statement

Original source: https://www.janestreet.com/puzzles/expelled/

There is no problem description this time, just the following two tables:

![Sample](expelled_.jpg "Sample triangle")

## Solution

It looks like we have to find the number that goes where the three question marks are in the second table.

Initially, it is not entirely clear how the two tables relate. I decided to focus on the first table first to see if there is any pattern in there. Obviously, part of the pattern is clear quite quickly:

![Sample](expelled_1.png "Pattern")

The yellow part of the grid can be generated by writing the numbers 1...n in row 1, 2...n in row 2, 3...n in row 3 etc. and then removing 1 cell from the left in the 3rd row, 3 cells from the left in the 4th row, 5 cells from the left in the 5th row , 7 cells from the left in the 6th row etc.

This alone, doesn't really bring us any further. Next, I decided to use some colors on the board and this is what explained what was going on:

![Sample](expelled_2.png "Pattern")

It is not entirely simple to see, but I noticed that in every row, the cell in the red square in the previous row is eliminated (expelled) and then the number closest to the left of the square is added. After that the number closest to the right side is added and so on. Furthermore, when a number is elminated, it is noted in the second table. For example: number 3 is elminated in the 2nd row, 4 is elminated in the 5th row, 5 is eliminated in the 3rd row, 6 is elminiated in the 9th row etc.

![Sample](expelled_3.png "Pattern")

![Sample](expelled_4.png "Pattern")

This means we can implement a simple algorithm that will return the rows and will also return a dictionary where the key is the number that was eliminated and the value of the dictionary is the iteration in which the key is eliminated.

In [1]:
def iterate(iterations, longest_row, print_rows):
    last_row = list(range(1,longest_row)) # initialize the first row

    if print_rows: 
        print('  '.join(['%2d'%c for c in last_row]))
        
    expelled_dict = dict()

    for row in range(2,iterations):
        next_row = []
        expelled = last_row[row-2]
        expelled_dict[expelled] = row-1
        pre_expelled = last_row[:row-2]
        post_expelled = last_row[row-1:]
        while len(pre_expelled) > 0:
            next_row.append(pre_expelled.pop())
            next_row.append(post_expelled.pop(0))
        next_row += post_expelled
        if print_rows: 
            print('  '.join(['%2d'%c for c in next_row]))
        last_row = next_row

    return expelled_dict

With this method, let's try to reproduce the initial grid:

In [2]:
iterate(10, 27, True)

 1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26
 2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26
 2   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26
 4   6   2   7   8   9  10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26
 2   8   6   9   4  10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26
 9  10   6  11   8  12   2  13  14  15  16  17  18  19  20  21  22  23  24  25  26
 8   2  11  13   6  14  10  15   9  16  17  18  19  20  21  22  23  24  25  26
14  15   6   9  13  16  11  17   2  18   8  19  20  21  22  23  24  25  26
11   2  16  18  13   8   9  19   6  20  15  21  14  22  23  24  25  26


{1: 1, 3: 2, 5: 3, 7: 4, 4: 5, 12: 6, 10: 7, 17: 8}

At this point, we don't have the result for number 11 yet. We need to run this with a longer grid and with more iterations and then we can check the expelled_dict variable:

In [3]:
expelled_dict = iterate(2000, 10000, False)

# When is number 2 expelled?
print('Number 2 is expelled after %i iterations: '%expelled_dict[2])

# When is number 11 expelled?
print('Number 11 is expelled after %i iterations: '%expelled_dict[11])

Number 2 is expelled after 75 iterations: 
Number 11 is expelled after 416 iterations: 


That solves the puzzle. Number 11 is expelled after 416 iterations.